File: internal/integrations/hermes/manifest.go
package hermes

import (
	"bytes"

	"gopkg.in/yaml.v3"
)

// Manifest 对应 manifest v2 的完整字段视图。
// v2 新增顶层 routing 映射（模型别名 → 实际模型名）和
// resources.skills 技能包列表；org/app 层规则已移出，仅保留平台规则。
// 字段顺序通过显式 yaml tag 控制；不引入 schema_version。
type Manifest struct {
	App         ManifestApp         `yaml:"app"`
	Credentials ManifestCredentials `yaml:"credentials"`
	Resources   ManifestResources   `yaml:"resources"`
	// Knowledge 是 manager runtime API 配置；Hermes 只拿 manager app token，不拿 RAGFlow 凭证。
	Knowledge ManifestKnowledge `yaml:"knowledge,omitempty"`
	// WebPublish 是 oc-publish skill 的运行时配置；企业未开通发布能力时整段省略。
	WebPublish ManifestWebPublish `yaml:"web_publish,omitempty"`
	// Routing 智能路由映射，键为模型别名，值为实际模型名；空时省略。
	Routing map[string]string `yaml:"routing,omitempty"`
	// Capabilities 是 manager 显式授予本次运行时的能力上限；空时表示沿用普通应用兼容语义。
	Capabilities []string `yaml:"capabilities,omitempty"`
}

// ManifestApp 业务元数据。id/name 仅审计日志使用；model 直接进 hermes config.yaml model.default。
// Language 是 hermes bot 对终端用户说话的语言（en/zh）；空时 renderer 回退平台默认，omitempty 省略。
type ManifestApp struct {
	ID       string `yaml:"id"`
	Name     string `yaml:"name"`
	Model    string `yaml:"model"`
	Language string `yaml:"language,omitempty"`
}

// ManifestCredentials 凭证集合；当前仅 openai；微信凭证由 hermes 自管。
type ManifestCredentials struct {
	OpenAI ManifestOpenAI `yaml:"openai"`
}

// ManifestOpenAI OPENAI 凭证；base_url 不带 /v1，由镜像 renderer 自行拼。
type ManifestOpenAI struct {
	APIKey  string `yaml:"api_key"`
	BaseURL string `yaml:"base_url"`
}

// ManifestResources 指向 resources/ 子目录的相对路径集合。
// v2 新增 skills 字段，指向已推送到 input/ 的技能包 tar 相对路径列表；空时省略。
type ManifestResources struct {
	Persona string        `yaml:"persona"`
	Rules   ManifestRules `yaml:"rules"`
	// Skills 技能包相对路径列表，例如 ["resources/skills/weather.tar"]；空时省略。
	Skills []string `yaml:"skills,omitempty"`
}

// ManifestRules 平台层规则文件的相对路径。
// v2 仅保留 platform 一层；org/app 层规则已由版本实例 system_prompt 覆盖。
type ManifestRules struct {
	Platform string `yaml:"platform"`
}

// ManifestKnowledge 描述 Hermes 通过 manager runtime API 使用知识库所需的最小配置。
type ManifestKnowledge struct {
	RuntimeBaseURL string `yaml:"runtime_base_url"`
	AppToken       string `yaml:"app_token"`
}

// ManifestWebPublish 是 oc-publish skill 的运行时配置（条件注入；企业未开通时整段省略）。
type ManifestWebPublish struct {
	RuntimeBaseURL string `yaml:"runtime_base_url"`
	AppToken       string `yaml:"app_token"`
	BaseDomain     string `yaml:"base_domain"`
}

// MarshalManifestYAML 把 Manifest 序列化为 YAML。
// 显式构造 yaml.Encoder 是为了未来需要时方便加 SetIndent 等。
func MarshalManifestYAML(m Manifest) ([]byte, error) {
	var buf bytes.Buffer
	enc := yaml.NewEncoder(&buf)
	enc.SetIndent(2)
	if err := enc.Encode(m); err != nil {
		return nil, err
	}
	if err := enc.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

File: runtime/hermes/hermes-v0.18.2/lib/manifest.py
"""manager 写入的 manifest.yaml 解析与必填字段校验。

仅暴露 forward-compat 的字段视图：调用方拿到 Manifest 实例后访问命名属性，
未知字段被忽略；缺必填字段则抛 ManifestError，让 oc-entrypoint 直接退出 1。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Union

import yaml


class ManifestError(Exception):
    """manifest 缺失必填字段或解析失败。"""


@dataclass(frozen=True)
class Manifest:
    """业务化视图。字段语义对应 spec §4.2 / §5（manifest v2）。"""
    app_id: str
    app_name: str
    app_model: str
    openai_api_key: str
    openai_base_url: str
    persona_rel: str
    rule_platform_rel: str
    # 组织层 / 应用层 rule：manifest v2 不再写；解析为可选，缺省空串。
    rule_organization_rel: str = ""
    rule_application_rel: str = ""
    # routing：8 个 auxiliary 槽位到模型名的映射；缺省空 dict（全部走主模型）。
    routing: dict = field(default_factory=dict)
    # skills：版本 skill tar 的相对路径列表（相对 input_root）；缺省空 list。
    skills: list = field(default_factory=list)
    # knowledge：manager runtime API 配置；不包含 RAGFlow 凭证。
    knowledge_runtime_base_url: str = ""
    knowledge_app_token: str = ""
    # web_publish：静态站点发布 API 配置；不含时表示企业未开通发布能力。
    web_publish_runtime_base_url: str = ""
    web_publish_app_token: str = ""
    web_publish_base_domain: str = ""
    # app_language：应用界面语言，由 manager 写入 manifest（"en"/"zh"）；
    # 缺省空串表示未配置，渲染时回落 "en"。
    app_language: str = ""


def _require(d: dict, *path: str) -> Any:
    """逐层取值；任一层缺失抛 ManifestError，错误信息指明字段路径。"""
    cur: Any = d
    for k in path:
        if not isinstance(cur, dict) or k not in cur:
            raise ManifestError(f"manifest missing field: {'.'.join(path)}")
        cur = cur[k]
    if cur in (None, ""):
        raise ManifestError(f"manifest empty field: {'.'.join(path)}")
    return cur


def load(path: Union[str, Path]) -> Manifest:
    """读 manifest.yaml 并构造 Manifest；未知顶层字段忽略；routing / skills / 组织层 / 应用层 rule 可选。"""
    raw = yaml.safe_load(Path(path).read_text())
    if not isinstance(raw, dict):
        raise ManifestError("manifest yaml root must be a mapping")
    resources = raw.get("resources")
    rules = resources.get("rules") if isinstance(resources, dict) else None
    rules = rules if isinstance(rules, dict) else {}
    routing = raw.get("routing")
    skills = resources.get("skills") if isinstance(resources, dict) else None
    knowledge = raw.get("knowledge")
    knowledge = knowledge if isinstance(knowledge, dict) else {}
    # web_publish：可选段，缺失或非 dict 时视为空配置（企业未开通发布能力）。
    wp = raw.get("web_publish")
    wp = wp if isinstance(wp, dict) else {}
    # 从 app 节点读取可选字段 language；不存在或为空时缺省空串，渲染侧回落 "en"。
    app_section = raw.get("app") if isinstance(raw.get("app"), dict) else {}
    return Manifest(
        app_id=_require(raw, "app", "id"),
        app_name=_require(raw, "app", "name"),
        app_model=_require(raw, "app", "model"),
        openai_api_key=_require(raw, "credentials", "openai", "api_key"),
        openai_base_url=_require(raw, "credentials", "openai", "base_url"),
        persona_rel=_require(raw, "resources", "persona"),
        rule_platform_rel=_require(raw, "resources", "rules", "platform"),
        rule_organization_rel=str(rules.get("organization") or ""),
        rule_application_rel=str(rules.get("application") or ""),
        routing={str(k): str(v) for k, v in routing.items()} if isinstance(routing, dict) else {},
        skills=[str(s) for s in skills] if isinstance(skills, list) else [],
        knowledge_runtime_base_url=str(knowledge.get("runtime_base_url") or ""),
        knowledge_app_token=str(knowledge.get("app_token") or ""),
        web_publish_runtime_base_url=str(wp.get("runtime_base_url") or ""),
        web_publish_app_token=str(wp.get("app_token") or ""),
        web_publish_base_domain=str(wp.get("base_domain") or ""),
        app_language=str(app_section.get("language") or ""),
    )
